proto/packet.go

```diff
@@ -175,61 +212,8 @@ func (cmd *Packet) Payload() (interface{}, error) {
 		return &ErrorReply{Error: cmd.Error}, nil
 	}
 
-	var payload interface{}
-
-	// TODO: use reflect + a map
-	switch cmd.Type {
-	case SendType:
-		payload = &SendCommand{}
-	case SendReplyType:
-		payload = &SendReply{}
-	case SendEventType:
-		payload = &SendEvent{}
-	case EditMessageType:
-		payload = &EditMessageCommand{}
-	case EditMessageEventType:
-		payload = &EditMessageEvent{}
-	case EditMessageReplyType:
-		payload = &EditMessageReply{}
-	case LogType:
-		payload = &LogCommand{}
-	case LogEventType:
-		payload = &LogEvent{}
-	case LogReplyType:
-		payload = &LogReply{}
-	case JoinEventType, PartEventType:
-		payload = &PresenceEvent{}
-	case NickType:
-		payload = &NickCommand{}
-	case NickReplyType:
-		payload = &NickReply{}
-	case NickEventType:
-		payload = &NickEvent{}
-	case PingType:
-		payload = &PingCommand{}
-	case PingEventType:
-		payload = &PingEvent{}
-	case PingReplyType:
-		payload = &PingReply{}
-	case AuthType:
-		payload = &AuthCommand{}
-	case AuthEventType:
-		payload = &AuthEvent{}
-	case AuthReplyType:
-		payload = &AuthReply{}
-	case BounceEventType:
-		payload = &BounceEvent{}
-	case NetworkEventType:
-		payload = &NetworkEvent{}
-	case SnapshotEventType:
-		payload = &SnapshotEvent{}
-	case WhoType:
-		return &WhoCommand{}, nil
-	case WhoEventType:
-		payload = &WhoEvent{}
-	case WhoReplyType:
-		payload = &WhoReply{}
-	default:
+	payload, ok := payloadMap[cmd.Type]
```

Comment:
This will return a pointer to the same value each time. This is where the reflect package needs to come in.

Change payloadMap from map[PacketType]interface{} to map[PacketType]reflect.Type. You can then use something like:

```
payloadType, ok := payloadMap[cmd.Type]
if !ok {
  // return error
}
return reflect.New(payloadType).Interface()
```